README.md
pyqrcode
================================

The pyqrcode module is a QR code generator that is simple to use and written
in pure python. The module can automates most of the building process for
creating QR codes. Most codes can be created using only two lines of code!

Unlike other generators, all of the helpers can be controlled manually. You are
free to set any or all of the properties of your QR code.

QR codes can be saved as SVG, XBM, EPS, PNG (by using the
[pypng](https://pypi.python.org/pypi/pypng/) module), or plain text. They can
also be displayed directly in most Linux terminal emulators and Tkinter. PIL
is not used to render the image files.

The pyqrcode module attempts to follow the QR code standard as closely as
possible. The terminology and the encodings used in pyqrcode come directly
from the standard. This module also follows the algorithm laid out in the
standard.

Requirements
-------------------------

The pyqrcode module only requires Python 2.6, Python 2.7, or Python 3. You may
want to install pypng in order to render PNG files, but it is optional.

Installation
------------

Installation is simple. It can be installed from pip using the following
command:

```bash
$ pip install pyqrcode
```

Or from the code

```bash
$ python setup.py install
```

Usage
-----

This is the only import you need. The heart of the module is the QRCode class.
You can construct the class normally, or use the *create* wrapper function.

```python
>>> import pyqrcode
>>> qr = pyqrcode.create('Unladden swallow')
>>> qr.png('famous-joke.png', scale=5)
```

PyPi
----

* _PyPi page_: https://pypi.python.org/pypi?name=PyQRCode&:action=display

* _Documentation_: http://pythonhosted.org/PyQRCode/

### Encoding Data ###

This module supports all four encodings for data: numeric, alphanumeric, kanji,
and binary.

The numeric type is the most efficient way to encode digits. As the
name implies it is designed to encode integers. Some numbers might be two
large, the object can use a string containing only digits instead of an
actual number.

```python
>>> number = pyqrcode.create(123456789012345)
````

The alphanumeric type is very limited in that it can only encode some ASCII
characters. It encodes: uppercase letters, 0-9, the horizontal space, and eight
punctuation characters. The available characters will let you encode a URL 

```python
>>> url = pyqrcode.create('http://uca.edu')
```

When all else fails the data can be encoded in pure binary. The quotation below
must be encoded in binary because of the lower-cased characters, the apostrophe
and the new line character.

```python
>>> life = pyqrcode.create('''MR. CREOSOTE: Better get a bucket. I'm going to throw up.
    MAITRE D: Uh, Gaston! A bucket for monsieur. There you are, monsieur.''')
```
The only unimplemented encoding is ECI mode which allows for multiple encodings in one QR
code (this will be implemented in a future version).

### Manually Setting The QR Code's Properties ###

There are many situation where you might wish to have more fine grained control
over how the QR Code is generated. You can specify all the properties of your
QR code through the *create* function. There are three main properties to a
QR code.

The _error_ parameter sets the error correction level of the code. Each level
has an associated name given by a letter: L, M, Q, or H; each level can
correct up to 7, 15, 25, or 30 percent of the data respectively. There are
several ways to specify the level, see pyqrcode.tables.modes for all the
possible values. By default this parameter is set to 'H' which is the highest
possible error correction, but it has the smallest available data
capacity.

The _version_ parameter specifies the size and data capacity of the
code. Versions are any integer between 1 and 40, where version 1 is
the smallest QR code, and version 40 is the largest. By default, the object
uses the data's encoding and error correction level to calculate the smallest
possible version. You may want to specify this parameter for consistency when
generating several QR codes with varying amounts of data. That way all of the
generated codes would have the same size.

Finally, the _mode_ parameter sets how the contents will be encoded. As
mentioned above, three of the five possible encodings have been written. By
default the object uses the most efficient encoding for the contents. You can
change this though. See qrcode.tables.modes for a list of possible values
for this parameter.

The code below constructs a QR code with 25% error correction, size 27, and
forces the encoding to be binary (rather than numeric).

```python
>>> big_code = pyqrcode.create('0987654321', error='L', version=27, mode='binary')
```

### Rendering ###

There are many possible formats for rendering the QR Code. The first is
to render it as a string of 1's and 0's. This is method is used to help end
users create their own renderer. It is also possible to print the
code such that it is directly displayable in most Linux terminals.
There are several image based renderers.

The terminal renderer outputs a string of ASCII escape codes that when
displayed in a compatible terminal, will display a valid QR code. The
background and module colors are settable (although as with any time you display
colors in the terminal, there are several caveats).

```python
>>> print(url.terminal())
>>> print(url.terminal('red', 'white'))
```

The SVG renderer outputs the QR Code as a scalable vector graphic. This
renderer does not require any external modules. Instead it hand draws the
QR code as a set paths.

```python
>>> url.svg(sys.stdout, scale=1)
>>> url.svg('uca.svg', scale=4, module_color="#7D007D")
```

Alternatively, if you install the pypng module, you can render the QR Code
to a PNG file. Colors should be specified as RGB or RGBA if you want to
take advantage of transparency.

```python
>>> number.png('big-number.png')
>>> life.png('sketch.png', scale=6, module_color=(0, 0, 0, 128), background=(0xff, 0xff, 0xcc))
```

Finally, there is a text based renderer. This will output the QR code as a
string of 1's and 0's, with each row of the code on a new line.

```python
>>> print(number.text())
```


Changelog
# 1.2.2
* Added `qrprint` as a CLI application to print QR codes to the terminal.

# 1.2.1
* Fixed issue #43. A debug print statement got left in by mistake. I altered
  The distribution script to check and make sure it does not happen again.

# 1.2
* Added Kanji support.
* Added ability to output PNG QR codes as a base64 string. Allows coded to be
  created for web services without the need to create intermediary files.
  Thanks to [Fábio C. Barrionuevo da Luz (luzfcb)](https://github.com/luzfcb)
* Added renderer for XBM. Displaying QR codes in Tkinter is now extremely
  simple. Thanks to [Seth VanHeulen (svanheulen)](https://github.com/svanheulen)

# 1.1.1
* Fix for issue #38, where numeric encodings got broken by added unicode support.

# 1.1
* Added support for Python 2.6
* All renderers now have a **quiet zone of four**. This value is settable via a
  parameter.
* Fixed issue where file streams were not being closed correctly
* **Special thanks goes to [Lars (heuer)](https://github.com/heuer) who 
  contributed a massive amount of improvements in this version.**
  * Enormously improved SVG implementation. Now uses paths instead of lines.
    Also allows for SVG fragments instead of entire documents.
  * We now have unit tests!! He wrote over 100 unit tests for
    all of the various parts of the library.
  * A new EPS renderer.
  * A mechanism for showing QR codes directly from within your code.

# 1.0
* Fixed issue where terminal bits were being added in the wrong location.
* Added ability to output QR code to a Linux terminal.
* Added support for Python 2.7


__init__.py
import pyqrcode.tables
import pyqrcode.builder as builder

class QRCode:
    """This class represents a QR code. To use this class simply give the
    constructor a string representing the data to be encoded, it will then
    build a code in memory. You can then save it in various formats. Note,
    codes can be written out as PNG files but this requires the PyPNG module.
    You can find the PyPNG module at http://packages.python.org/pypng/.
    
    Examples:
        >>> from pyqrcode import QRCode
        >>> import sys
        >>> url = QRCode('http://uca.edu')
        >>> url.svg(sys.stdout, scale=1)
        >>> url.svg('uca.svg', scale=4)
        >>> number = QRCode(123456789012345)
        >>> number.png('big-number.png')
    """
    def __init__(self, content, error='H', version=None, mode=None):
        """When creating a QR code only the content to be encoded is required,
        all the other properties of the code will be guessed based on the
        contents given. When the QRCode object is created the QR code will be
        generated immediately.
        
        Unless you are familiar with QR code's inner workings 
        it is recommended that you just specify the content and nothing else.
        However, there are cases where you may want to specify the various
        properties of the created code manually, this is what the other
        parameters do. Below, you will find a lengthy explanation of what
        each parameter is for. Note, the parameter names and values are taken
        directly from the standards. You may need to familiarize yourself
        with the terminology of QR codes for the names to make sense.
        
        The error parameter sets the error correction level of the code. There
        are four levels defined by the standard. The first is level 'L' which
        allows for 7% of the code to be corrected. Second, is level 'M' which
        allows for 15% of the code to be corrected. Next, is level 'Q' which
        is the most common choice for error correction, it allow 25% of the
        code to be corrected. Finally, there is the highest level 'H' which
        allows for 30% of the code to be corrected. There are several ways to
        specify this parameter, you can use an upper or lower case letter,
        a float corresponding to the percentage of correction, or a string
        containing the percentage. See tables.modes for all the possible
        values. By default this parameter is set to 'H' which is the highest
        possible error correction, but it has the smallest available data
        capacity.
        
        The version parameter specifies the size and data capacity of the
        code. Versions are any integer between 1 and 40. Where version 1 is
        the smallest QR code, and version 40 is the largest. If this parameter
        is left unspecified, then the contents and error correction level will
        be used to guess the smallest possible QR code version that the 
        content will fit inside of. You may want to specify this parameter
        for consistency when generating several QR codes with varying amounts
        of data. That way all of the generated codes would have the same size.
        
        The mode parameter specifies how the contents will be encoded. By
        default, the best possible encoding for the contents is guessed. There
        are four possible encoding methods. First, is 'numeric' which is
        used to encode integer numbers. Next, is 'alphanumeric' which is
        used to encode some ASCII characters. This mode uses only a limited
        set of characters. Most problematic is that it can only use upper case
        English characters, consequently, the content parameter will be
        subjected to str.upper() before encoding. See tables.ascii_codes for
        a complete list of available characters. We then have 'binary' encoding
        which just encodes the bytes directly into the QR code (this encoding
        is the least efficient). Finally, there is 'kanji'  encoding (i.e.
        Japanese characters), this is unimplemented at this time.
        """
        #Coerce the content into a string
        self.data = str(content)
        
        #Check that the passed in error level is valid
        try:
            self.error = tables.error_level[str(error).upper()]
        except:
            raise ValueError('The error parameter is not one of '
                             '"L", "M", "Q", or "H."')
        
        #Guess the mode of the code, this will also be used for
        #error checking
        guessed_content_type = self._detect_content_type()
        
        #Force a passed in mode to be lowercase
        if mode:
            mode = mode.lower()
        
        #Check that the mode parameter is compatible with the contents
        if not mode:
            #Use the guessed mode
            self.mode = guessed_content_type
            self.mode_num = tables.modes[self.mode]
        elif guessed_content_type == 'binary' and \
             tables.modes[mode] != tables.modes['binary']:
            #Binary is only guessed as a last resort, if the
            #passed in mode is not binary the data won't encode
            raise ValueError('The content provided cannot be encoded with '
                             'the mode {}, it can only be encoded as '
                             'binary.'.format(mode))
        elif tables.modes[mode] == tables.modes['numeric'] and \
             guessed_content_type != 'numeric':
            #If numeric encoding is requested make sure the data can
            #be encoded in that format
            raise ValueError('The content cannot be encoded as numeric.')
        else:
            #The data should encode with the passed in mode
            self.mode = mode
            self.mode_num = tables.modes[self.mode]

        #Guess the "best" version
        self.version = self._pick_best_fit()
        
        #If the user supplied a version, then check that it has
        #sufficient data capacity for the contents passed in            
        if version:
            if version >= self.version:
                self.version = version
            else:
                raise ValueError('The data will not fit inside a version {} '
                                 'code with the given encoding and error '
                                 'level (the code must be at least a '
                                 'version {}).'.format(version, self.version))
        
        #Build the QR code
        self.builder = builder.QRCodeBuilder(data=content,
                                     version=self.version,
                                     mode=self.mode,
                                     error=self.error)
       
        #Save the code for easier reference
        self.code = self.builder.code

    def _detect_content_type(self):
        """This method tries to auto-detect the type of the data. It first
        tries to see if the data is a valid integer, in which case it returns
        numeric. Next, it tests the data to see if it is 'alphanumeric.' QR
        Codes use a special table with very limited range of ASCII characters.
        The code's data is tested to make sure it fits inside this limited
        range. If all else fails, the data is determined to be of type
        'binary.'
        
        Note, encoding 'kanji' is not yet implemented.
        """
        #See if the data is an integer
        try:
            test = int(content)
            return 'numeric'
        except:
            #Content is not numeric, this is not an error
            pass
        
        #See if that data is alphanumeric based on the standards
        #special ASCII table
        valid_characters = tables.ascii_codes.keys()
        if all(map(lambda x: x in valid_characters, self.data.upper())):
            return 'alphanumeric'
           
        #All of the tests failed. The content can only be binary. 
        return 'binary'

    def _pick_best_fit(self):
        """This method return the smallest possible QR code version number
        that will fit the specified data with the given error level.
        """
        for version in range(1,41):
            #Get the maximum possible capacity
            capacity = tables.data_capacity[version][self.error][self.mode_num]
            
            #Check the capacity
            if (self.mode_num == tables.modes['binary'] and \
               capacity >= len(self.data.encode('ascii'))) or \
               capacity >= len(self.data):
                return version
        
        raise ValueError('The data will not fit in any QR code version '
                         'with the given encoding and error level.')
    
    def get_png_size(self, scale):
        """This is method helps users determine what scale to use when creating
        a PNG of this QR code. It is meant mostly to be used in the console
        to help a user figure out what scale to use. It will return an integer
        representing the width and height of the QR code in pixels, if it was
        drawn using the given scale.
        
        Example:
            >>> code = pyqrcode.QRCode("I don't like spam!")
            >>> print(code.get_png_size(1))
            31
            >>> print(code.get_png_size(5))
            155
        """
        return builder._get_png_size(self.version, scale)
        
    def png(self, file, scale=1, module_color=None, background=None):
        """This method writes the QR code out as an PNG image. The resulting
        PNG has a bit depth of 1. The file parameter is used to specify where
        to write the image to. It can either be an writable stream or a
        file path.
        
        The scale parameter sets how large to draw a single module. By
        default one pixel is used to draw a single module. This may make the
        code to small to be read efficiently. Increasing the scale will make
        the code larger. Only integer scales are usable. This method will
        attempt to coerce parameter into an integer (e.g. 2.5 will become 2,
        and '3' will become 3).
        
        The module_color parameter sets what color to use for the encoded
        modules (the black part on most QR codes). The background parameter
        sets what color to use for the background (the white part on most
        QR codes). If either parameter is set, then both must be
        set or a ValueError is raised. Colors should be specified as either
        a list or a tuple of length 3 or 4. The components of the list must
        be integers between 0 and 255. The first three member give the RGB
        color. The fourth member gives the alpha component, where 0 is
        transparent and 255 is without transparency. Note, many color
        combinations are unreadable by scanners, so be careful.
        
        This method will write the given file out as a PNG file. Note, it
        depends on the PyPNG module to do this.
        
        Example:
            >>> code = pyqrcode.QRCode('Are you suggesting coconuts migrate?')
            >>> code.png('swallow.png', scale=5)
            >>> code.png('swallow.png', scale=5,
                         module_color=(0x66, 0x33, 0x0),      #Dark brown
                         background=(0xff, 0xff, 0xff, 0x88)) #50% transparent white
        """
        builder._png(self.code, self.version, file, scale,
                     module_color, background)
    
    def svg(self, file, scale=1, module_color='#000000', background=None):
        """This method writes the QR code out as an SVG document. The
        code is drawn by drawing only the modules corresponding to a 1. They
        are drawn using a line, such that contiguous modules in a row
        are drawn with a single line. The file parameter is used to
        specify where to write the document to. It can either be a writable
        stream or a file path. The scale parameter is sets how large to draw
        a single module. By default one pixel is used to draw a single
        module. This may make the code to small to be read efficiently.
        Increasing the scale will make the code larger. Unlike the png() method,
        this will accept fractional scales (e.g. 2.5).
        
        Note, three things are done to make the code more appropriate for
        embedding in a HTML document. The "white" part of the code is actually
        transparent. The code itself has a class of "pyqrcode". The lines
        making up the QR code have a class pyqrline. These should make the code
        easier to style using CSS.
        
        You can also set the colors directly using the module_color and
        background parameters. The module_color parameter sets what color to
        use for the encoded modules (the black part on most QR codes). The
        background parameter sets what color to use for the background (the
        white part on most QR codes). The parameters can be set to any valid
        SVG or HTML color. If the background is set to None, then no background
        will be drawn, i.e. the background will be transparent. Note, many color
        combinations are unreadable by scanners, so be careful.
        
        Example:
            >>> code = pyqrcode.QRCode('Hello. Uhh, can we have your liver?')
            >>> code.svg('live-organ-transplants.svg', 3)
        """
        builder._svg(self.code, self.version, file, scale,
                     module_color, background)
        
    def text(self):
        """This method returns a string based representation of the QR code.
        The black modules are represented by 1's and the white modules are
        represented by 0's. This is useful for debugging purposes. It can also
        be used to allow a user to write their own output function.
        """
        return builder._text(self.code)

if __name__ == '__main__':
    code = QRCode('''This is an example!!''')
    code.png( "best-code.png", scale=4)

